org.eclipse.xtext/src/org/eclipse/xtext/xtext/XtextValidator.java

```diff
@@ -1251,27 +1252,26 @@ public void checkOppositeReferenceUsed(Assignment assignment) {
 	@Check
 	public void checkOverridingRule(AbstractRule rule) {
 		final String name = rule.getName();
-		
-		final boolean isOverride =
-				rule.getAnnotations().stream().anyMatch(e -> AnnotationNames.OVERRIDE.equals(e.getName()));
-		
 		final List<Grammar> superGrammars = GrammarUtil.getGrammar(rule).getUsedGrammars();
-		
+		boolean isOverride = isOverride(rule);
 		if (isOverride && superGrammars.isEmpty()) {
 			error("This grammar has no super grammar and therefore cannot override any rules.", rule,
 					XtextPackage.Literals.ABSTRACT_RULE__NAME, XtextConfigurableIssueCodes.EXPLICIT_OVERRIDE_INVALID);
 		}
 		
 		for (Grammar g : superGrammars) {
 			final AbstractRule r = GrammarUtil.findRuleForName(g, rule.getName());
-			
+
 			if (r != null) {
-				if (!isOverride) {
+				if(isFinal(r)) {
+					error("This rule illegally overrides " + name + " in " + GrammarUtil.getGrammar(r).getName() + " which is final.", rule, 
```

Comment:
can and should this check deprecated too?

Reply:
We could, right.